src/compiler/factory.ts

```diff
@@ -3055,4 +3055,539 @@ namespace ts {
+        //   param[_i - restIndex] = arguments[_i];
+        // }
+        const forStatement = createFor(
+            createVariableDeclarationList([
+                createVariableDeclaration(temp, /*type*/ undefined, createLiteral(restIndex))
+            ], /*location*/ parameter),
+            createLessThan(
+                temp,
+                createPropertyAccess(createIdentifier("arguments"), "length"),
+                /*location*/ parameter
+            ),
+            createPostfixIncrement(temp, /*location*/ parameter),
+            createBlock([
+                startOnNewLine(
+                    createStatement(
+                        createAssignment(
+                            createElementAccess(
+                                expressionName,
+                                createSubtract(temp, createLiteral(restIndex))
+                            ),
+                            createElementAccess(createIdentifier("arguments"), temp)
+                        ),
+                        /*location*/ parameter
+                    )
+                )
+            ])
+        );
+
+        setEmitFlags(forStatement, EmitFlags.CustomPrologue);
+        startOnNewLine(forStatement);
+        statements.push(forStatement);
+    }
+
+
+
+
+    export function convertForOf(node: ForOfStatement, convertedLoopBodyStatements: Statement[],
+                                 visitor: (node: Node) => VisitResult<Node>,
+                                 enableSubstitutionsForBlockScopedBindings: () => void,
+                                 context: TransformationContext,
+                                 convertObjectRest?: boolean): ForStatement | ForOfStatement {
+        // The following ES6 code:
+        //
+        //    for (let v of expr) { }
+        //
+        // should be emitted as
+        //
+        //    for (var _i = 0, _a = expr; _i < _a.length; _i++) {
+        //        var v = _a[_i];
+        //    }
+        //
+        // where _a and _i are temps emitted to capture the RHS and the counter,
+        // respectively.
+        // When the left hand side is an expression instead of a let declaration,
+        // the "let v" is not emitted.
+        // When the left hand side is a let/const, the v is renamed if there is
+        // another v in scope.
+        // Note that all assignments to the LHS are emitted in the body, including
+        // all destructuring.
+        // Note also that because an extra statement is needed to assign to the LHS,
+        // for-of bodies are always emitted as blocks.
+
+        const expression = visitNode(node.expression, visitor, isExpression);
+        const initializer = node.initializer;
+        const statements: Statement[] = [];
+
+        // In the case where the user wrote an identifier as the RHS, like this:
+        //
+        //     for (let v of arr) { }
+        //
+        // we don't want to emit a temporary variable for the RHS, just use it directly.

```

Comment:
the only changes to convertForOf are 
1. in creating `counter` and `elementAccess`
2. calling flattenParameterDestructuring with `convertObjectRest`
3. calling `createForOf` vs `createFor` depending on whether `convertObjectRest` is true.

Otherwise it's just moved from es2015